File: Barnes_Hut/Barnes_hut.cpp
/*------------Barnes_hut.cpp--------------------------------------------------//
*
* Purpose: We start with an octree, then we move onto an N-body simulator
*
*   Notes: We need a separate particle struct to hold pos, mass, radius
*          Depth-first-search
*          force integration
*          visualization
*
*-----------------------------------------------------------------------------*/

#include <iostream>
#include <cstdio> // For printf (which should replace iostream)
#include <vector>
#include <array>
#include <random>

/*----------------------------------------------------------------------------//
* STRUCTS
*-----------------------------------------------------------------------------*/

// Struct for positions of particles and octree nodes
struct pos{
    double x, y, z;
    pos() : x(0.0), y(0.0), z(0.0) {}
    pos(double x0, double y0, double z0) : x(x0), y(y0), z(z0) {}
};

// Struct for Center of mass
struct center_of_mass{
    pos p;
    double mass;
};

// struct for octree nodes
struct node {
    node *parent;

    // Position of node / box
    pos p;
    double box_length;
    center_of_mass com;

    // Positions of all particles in box
    std::vector<pos> p_pos;
 
    // Nodes are ordered in the following way:
    // Split cube into north / south, then into NSEW again,
    // 0: NNE, 1: NSE, 2: NSW, 3: NNW, 4: SNE, 5: SSE, 6: SSW, 7: SNW
    std::array<node *, 8> children;

    node() : p(pos()), box_length(1.0) {}
    node(pos loc, double length, node *par) : 
        p(loc), box_length(length), parent(par) {}

};

// Function to create random distribution of particles for Octree
std::vector<pos> create_rand_dist(double box_length, int pnum);

// Function to create octree from position data
// Initially, read in root node
void divide_octree(std::vector<pos> &p_pos, node *curr, int box_threshold,
                   double mass);

// Function to check whether a position is within a box
bool in_box(node *curr, pos p);

// Function to create 8 children node for octree
void make_octchild(node *curr);

// Function to find center of mass
// NOTE: some objects may have different masses, maybe read in center of masses
center_of_mass find_com(std::vector<pos> &p_pos, double mass);

/*----------------------------------------------------------------------------//
* MAIN
*-----------------------------------------------------------------------------*/

int main(){
    std::vector<pos> p_pos = create_rand_dist(1.0, 10);
    //std::vector<pos> p_pos(1);
    //p_pos[0].x = 0.25; p_pos[0].y = 0.25; p_pos[0].z = 0.25;

    std::cout << '\n' << '\n';

    // Creating root node
    node *root = new node();
    divide_octree(p_pos, root, 1, 0.1);

/*
    for (auto &p : p_pos){
        if (in_box(root, p)){
            std::cout << "found particle" << '\n';
        }
    }
*/

}

/*----------------------------------------------------------------------------//
* SUBROUTINES
*-----------------------------------------------------------------------------*/

// Function to create random distribution of particles for Octree
std::vector<pos> create_rand_dist(double box_length, int pnum){

    // Creating vector for particle positions (p_pos)
    std::vector<pos> p_pos;
    p_pos.reserve(pnum);

    // Creating random device to place particles later
    static std::random_device rd;
    int seed = rd();
    static std::mt19937 gen(seed);
    std::uniform_real_distribution<double> 
        box_dist(-box_length * 0.5, box_length * 0.5);

    std::cout << "creating random positions for all " << pnum
              << " particles!" << '\n';
    // Iterating over all particles to create random positions
    for (int i = 0; i < pnum; ++i){
        p_pos.emplace_back(box_dist(gen), box_dist(gen), box_dist(gen));
        printf("%lf %lf %lf \n", p_pos[i].x, p_pos[i].y, p_pos[i].z);
    }
    
    return p_pos;
}

// Function to create octree from position data
// Initially, read in root node
void divide_octree(std::vector<pos> &p_pos, node *curr, int box_threshold, 
                   double mass){
    // Divide node into 8 subnodes (boxes) to work with
    make_octchild(curr);
    
    // Iterating through all the children
    for (auto &child : curr->children){
        // Determine number of particles in our current box
        for (auto &p : p_pos){
            if (in_box(child, p)){
                child->p_pos.push_back(p);
                child->com.p.x += p.x;
                child->com.p.y += p.y;
                child->com.p.z += p.z;
                child->com.mass += 1;
                //child->com.mass +=p.mass;
            }
        }
        child->parent = curr;
        child->com.p.x /= child->com.mass;
        child->com.p.y /= child->com.mass;
        child->com.p.y /= child->com.mass;
        //std::cout << child->p_pos.size() << '\n';
        if (child->p_pos.size() > box_threshold){
            //child->com = find_com(child->p_pos, mass);
            std::cout << child->com.p.x << '\t' << child->com.p.y << '\t'
                      << child->com.mass << '\n';

            divide_octree(child->p_pos, child, box_threshold, mass);
        }
        else if(child->p_pos.size() <= box_threshold && 
                child->p_pos.size() != 0){
            //child->com = find_com(child->p_pos, mass);
            std::cout << child->com.p.x << '\t' << child->com.p.y << '\t'
                      << child->com.mass << '\n';
            std::cout << "Found particle!" << '\n';
        }
        //else{
        //    std::cout << "No particle in box!" << '\n';
        //}

    }

}

// Function to check whether a position is within a box
bool in_box(node *curr, pos p){
    return (p.x >= curr->p.x - curr->box_length * 0.5 && 
            p.x < curr->p.x + curr->box_length * 0.5 &&
            p.y >= curr->p.y - curr->box_length * 0.5 &&
            p.y < curr->p.y + curr->box_length * 0.5 &&
            p.z >= curr->p.z - curr->box_length * 0.5 &&
            p.z < curr->p.z + curr->box_length * 0.5 );
}

// Function to create 8 children node for octree
void make_octchild(node *curr){
    double node_length = curr->box_length * 0.5;

    // iterating through possible locations for new children nodes
    for (int k = -1; k <= 1; k+= 2){
        for (int j = -1; j <= 1; j += 2){
            for (int i = -1; i <= 1; i +=2){
                int n = 2 * k + j + (i+1)/2 + 3;
                curr->children[n] = new node();
                //std::cout << "n is: " << n << '\n';
                curr->children[n]->box_length = node_length;
                curr->children[n]->p.z = curr->p.z + k * curr->box_length*0.25;
                curr->children[n]->p.y = curr->p.y + j * curr->box_length*0.25;
                curr->children[n]->p.x = curr->p.x + i * curr->box_length*0.25;
            }
        }
    }
}

// Function to find center of mass
center_of_mass find_com(std::vector<pos> &p_pos, double mass){
    center_of_mass com;
    for (auto &p : p_pos){
        com.p.x += p.x;
        com.p.y += p.y;
        com.p.z += p.z;
    }

    // Assuming that all particles are of the same weight, this should be fine.
    com.p.x /= p_pos.size();
    com.p.y /= p_pos.size();
    com.p.z /= p_pos.size();

    com.mass = p_pos.size() * mass;

    return com;
}

File: Barnes_Hut/barnes_hut.jl
#-------------barnes_hut.jl----------------------------------------------------#
#
# Purpose: To implement the Barnes-Hut galaxy simulator in julia
#
#------------------------------------------------------------------------------#

# Type for particles later to be used with Verlet integration
type Particle
    p::Vector{Float64}
    prev_p::Vector{Float64}
    vel::Vector{Float64}
    acc::Vector{Float64}
    mass::Float64
    radius::Float64

    Particle(p::Vector{Float64}, prev_p::Vector{Float64}, vel::Vector{Float64}, 
             acc::Vector{Float64}, mass::Float64, radius::Float64) = 
        new(p, prev_p, vel, acc, mass, radius)
end

# type for node, to be used as an octree
type Node
    p::Vector{Float64}
    box_length::Float64
    parent::Nullable{Node}
    children::Vector{Node}
    com::Particle
    p_vec::Vector{Particle}
    Node(p::Vector{Float64}, box_length::Float64, parent::Nullable{Node}, 
         children::Vector{Node}, com::Particle, p_vec::Vector{Particle}) =
        new(p, box_length, parent, children, com, p_vec)
end

#------------------------------------------------------------------------------#
# SUBROUTINES
#------------------------------------------------------------------------------@

# Function to create random distribution
function create_rand_dist(box_length::Float64, pnum::Int)
    # initialize p_vec
    p_vec = fill(Particle(zeros(3), zeros(3), zeros(3), zeros(3),1.0,1.0), 0)
    for i = 1:pnum
        push!(p_vec, Particle(rand(3)*box_length-(box_length * 0.5),
                              rand(3)*box_length-(box_length * 0.5),
                              rand(3)*box_length-(box_length * 0.5),
                              rand(3)*box_length-(box_length * 0.5),
                              1, 1))
    end
    return p_vec
end

# Function to initialize root node of octree
function make_octree(p_vec::Vector{Particle}, box_length::Float64)
    com = Particle(zeros(3), zeros(3), zeros(3), zeros(3),1.0,1.0)
    root = Node([0.0; 0.0; 0.0], box_length, Nullable{Node}(), 
                Array(Node, 0), com, p_vec)
    return root
end

# Function to create children in a given node
function make_octchild(curr::Node)
    node_length = curr.box_length * 0.5
    quarter_box = curr.box_length * 0.25

    # Defining null com
    com = Particle(zeros(3), zeros(3), zeros(3), zeros(3),1.0,1.0)

    # iterating through possible location for new children nodes
    # wrtten by laurensbl
    for k = -1:2:1
        for j = -1:2:1
            for i = -1:2:1
                n = 2 * k + j + (i+1)/2 + 3;
                push!(curr.children, Node([curr.p[1] + k * quarter_box;
                                      curr.p[2] + j * quarter_box;
                                      curr.p[3] + i * quarter_box],
                                      curr.box_length*.5, Nullable{Node}(curr), 
                                      Array(Node, 0), com, 
                                      fill(com,0)))
            end
        end
    end
end

# Function to check if we are in a box
function in_box(curr::Node, p::Particle)
    half_box = curr.box_length * 0.5
    return p.p[1] >= curr.p[1] - half_box &&
           p.p[1] <  curr.p[1] + half_box &&
           p.p[2] >= curr.p[2] - half_box &&
           p.p[2] <  curr.p[2] + half_box &&
           p.p[3] >= curr.p[3] - half_box &&
           p.p[3] <  curr.p[3] + half_box

end

# Function to divide root node onto octree
function divide_octree(curr::Node, box_threshold::Int64)
    make_octchild(curr)
    for i = 1:8
        for p in curr.p_vec
            if (in_box(curr.children[i], p))
                push!(curr.children[i].p_vec, p)
                curr.children[i].p += p.p * p.mass
                curr.children[i].com.mass += p.mass
            end
        end
        curr.children[i].com.p /= curr.children[i].com.mass
        if (length(curr.children[i].p_vec) > box_threshold)
            divide_octree(curr.children[i], box_threshold)
        end
    end
end

# Function to output particle positions
# Note: Don't forget to opena nd close file surrounding this function
function particle_output(curr::Node, file::IOStream)
    if (length(curr.p_vec) == 1)
        writedlm(file,curr.p_vec[1].p')
    end

    for child in curr.children
        particle_output(child, file)
    end
end

# Function to find acceleration of single particle
function RKsearch(curr::Node, part::Particle)
    d = -part.p
    inverse_r = 1/sqrt(d[1]*d[1] + d[2]*d[2] + d[3]*d[3])

    if (inverse_r > 10.0)
        inverse_r = 10.0
    end
    if (inverse_r < -10.0)
        inverse_r = -10.0
    end

    part.acc += d * inverse_r * 10.0
    
end

# Verlet integration
function verlet(part::Particle, dt::Float64)
    temp = part.p
    part.p = 2 * part.p - part.prev_p + part.acc * dt * dt
    part.prev_p = temp
end

# Function to find acceleration of particles in Barnes-Hut tree
function force_integrate(root::Node, dt::Float64)

    # Going through all particles in p_vec to find a new acceleration
    for part in root.p_vec
        RKsearch(root, part)
    end

    # Going through all particles and updating position
    for part in root.p_vec
        verlet(part, dt)
    end
    
end

# Main function
function main()
    numsteps = 5
    p_output = open("pout_julia.dat", "w")

    vel = sqrt(10)
    dt = 0.01
    p_vec = fill(Particle([1.0; 0.0; 0.0],[cos(vel*dt); sin(vel*dt); 0.0],
                 [0.0; vel; 0.0],zeros(3),1E10,1.0), 1)
    println(p_vec[1].prev_p)

    root = make_octree(p_vec, 1000.0)
    divide_octree(root, 1)

    for i = 1:numsteps
        root = make_octree(p_vec, 1000.0)
        divide_octree(root, 1)
        if (i % 1 == 0 || (i-1) == 0)
            particle_output(root, p_output)
            write(p_output, "\n\n")
        end
        force_integrate(root, 0.01)
    end

    close(p_output)
end

@time main()
